### ISSUE TITLE
[eval] (unknown) error position with invalid Reflect.fields() call

### ISSUE THREAD

**Issue by Gama11** [HaxeFoundation/haxe#7750]
This code is obviously stupid, but shouldn't there be a proper position for the error message? `-D eval-stack` doesn't help either.

```haxe
class Main {
	static function main() {
		Reflect.fields(0);
	}
}
```

```
(unknown) : Uncaught exception Unexpected value VInt32(0), expected object
```

**Comment by Simn** [comment]
I still don't have a good strategy for this when it comes to builtin functions. We could probably use the position of the call but I don't think we could make it error on the `0` itself.

What position did this show in 3.4?

**Comment by Gama11** [comment]
The call seems good enough.

Well... it didn't error at all, it returned an empty array. :)

**Comment by Simn** [comment]
The problem with accurate error reporting on built-in functions is that I must absolutely avoid hitting OCaml's write-barrier. This means that I don't actually have a frame for a built-in function and that makes call-stack-management more tricky.

The current approach is this: Each frame carries `leave_pmin` and `leave_pmax` record fields which are updated when a call is made (they are ints so no write-barrier). When we build an exception stack, we ignore the topmost frame because its leave-position is by definition not relevant (else it wouldn't be the topmost frame).

This works really nicely, except for when built-in functions get involved because they don't push frames. In this case we _don't_ want to ignore the topmost frame because it is the one that called the built-in function.

This would mean that we would have to track if we're in a built-in function or not and unfortunately there's no good place to do that. The only thing I can think of is to have a wrapper around each built-in VFunction which handles this, but I'm pretty sure that this would really degrade performance as well.

So tl;dr I don't have a solution for this at the moment and I cannot promise one for 4.0.

**Comment by Simn** [comment]
One heuristic that might work in these cases: If the exception position itself is `null_pos` (as is the case when decoding something from within a built-in function), we could still show the topmost frame position. This means we would get this:

```
Uncaught exception Null Access
source/Main.hx:3: characters 3-28 : Called from here
```

As I've said there is no chance of getting the position on the value that actually fails to decode. This would require associating values with positions which is not something I'm willing to do.

My concern is that this might trigger in some cases where we have a `null_pos` for other reasons. But then again that shouldn't really happen.

**Comment by Gama11** [comment]
That seems fair, and is still a big improvement. And even still points you to exactly the correct line in this example.

**Comment by ncannasse** [comment]
This is a real show stopper, needs to be addressed for 4.0, can we get a benchmark result of performance impact ?

**Comment by ncannasse** [comment]
Maybe I didn't read correctly :) Does your fix covers 99.9% of the cases ? In which cases it's still a problem ?

**Comment by Gama11** [comment]
The `Reflect.fields()` code example now gives:

```
Uncaught exception Unexpected value VInt32(0), expected object
source/Main.hx:3: characters 9-26 : Called from here
```

**Comment by Simn** [comment]
I think it covers 101% of the cases, i.e. it might report nonsense in some situations. But that would necessarily involve `null_pos` which _should_ not happen in normal code.

**Comment by Simn** [comment]
So unless you find an actual problem I'll keep this closed.

Plus you can always press F5 if you need more information about macro errors. \o/
